README.md
# USB automator

[![Build Status](https://travis-ci.org/greizgh/usbautomator.svg?branch=master)](https://travis-ci.org/greizgh/usbautomator)

This little utility can run commands when a device is plugged or unplugged.

The main use case is to change keyboard layout according to device presence.
It can also be used to start a backup when your external drive is plugged.

## Configuration

On first run a default config file will be created in `$XDG_CONFIG_HOME/usbautomator/config.toml`.
Configuration follows this pattern:

```toml
[devices.NAME]
on_plugged = "COMMAND"
on_unplugged = "COMMAND"
[devices.NAME.properties]
ID_VENDOR_ID = "feed"
ID_MODEL_ID = "1307"
ID_INPUT_KEYBOARD = "1"
KEY = "1000000000007 ff9f207ac14057ff febeffdfffefffff fffffffffffffffe"
```

### How to configure an automation ?

Let's say you want to switch layout when your keyboard is plugged.
We will configure usbautomator step by step.

First, add a section with your commands:

```toml
[devices.mykeyboard]
on_plugged = "setxkbmap us"
on_unplugged = "setxkbmap <your laptop layout or whatever>"
```

You then need to find some properties which uniquely identify your device.
**Warning**: a single physical device might show up as several devices on your operating system.

You can use the *watch* command:

    usbautomator -w

This will listen for changes and display the properties of the devices (there might be a lot of them).
Simply plug your keyboard to list changed devices properties.
The command can be killed as usual with `<ctrl-c>`.

You should see this kind of properties:

```
ID_FOR_SEAT: input-pci-0000_00_14_0-usb-0_1_1_3
ID_INPUT: 1
ID_INPUT_KEY: 1
ID_INPUT_KEYBOARD: 1
ID_MODEL: ErgoDox_EZ
[...]
ID_MODEL_ID: 1307
ID_VENDOR: ErgoDox_EZ
ID_VENDOR_ID: feed
KEY: 1000000000007 ff9f007ac04007ff febeffdfffefffff fffffffffffffffe
NAME: "ErgoDox EZ ErgoDox EZ"
PRODUCT: 3/feed/1307/111
```

Some properties depend on the USB bus, those are not useful to identify your device as they might change depending on the USB plug used to connect your device.
The `KEY` property seems a good fit to identify your device, let's add it to the configuration:

```toml
[devices.mykeyboard.properties]
KEY = "1000000000007 ff9f007ac04007ff febeffdfffefffff fffffffffffffffe"
```

Your keyboard layout should now change according to the device presence.

**Note**: if the properties used in the configuration match multiple devices, the on_plugged/on_unplugged commands will run multiple time.

## Building

You will need [rust](https://www.rust-lang.org) (v1.31) and [cargo](https://doc.rust-lang.org/cargo/getting-started/installation.html).

Then simply run:

    cargo build --release

## Packages

[Archlinux](https://aur.archlinux.org/packages/usbautomator)

## Changelog

See [CHANGELOG.md](./CHANGELOG.md)

## License

GPLv3


CHANGELOG.md
# Changelog

## Unreleased
### Changed
- Make on_plugged/on_unplugged commands optional
- Switch to Rust 2018 edition
- Add nix shell config

## [0.3.1] - 2018-09-24
### Fixed
- Crashing when serializing the default config file
### Changed
- use cargo package name in notification instead of hardcoded value

## [0.3.0] - 2018-09-20
### Added
- Add option to list input and block devices
- Add option to describe device properties
- Allow to match device based on property set
- Command to watch for changes and display device properties
### Changed
- Updated sample config for ergodox
### Removed
- Restriction on subsystem type

## [0.2.0] - 2018-09-03
### Fixed
- Crashing when notify service is missing
- Do not crash when command fail

## [0.1.0] - 2018-08-08
### Added
- Trigger a command on device plug
- Trigger a command on device unplug
- Notify on device plug/unplug


src/main.rs
use libudev::{Context, Enumerator, Monitor};
use std::fs::File;
use std::io::prelude::*;
use std::thread::sleep;
use std::time::Duration;
use structopt::StructOpt;

mod config;
mod manager;

const SLEEP_DURATION: u64 = 1000;

#[derive(Debug, StructOpt)]
struct Opt {
    /// List connected devices of type input or block
    #[structopt(short = "l", long = "list")]
    list: bool,
    /// Watch for changes and display changed device properties
    #[structopt(short = "w", long = "watch")]
    watch: bool,
    /// List properties of device identified by given name
    #[structopt(long = "describe")]
    device_name: Option<String>,
}

fn main() {
    let opt = Opt::from_args();

    let config = get_config().unwrap();
    let manager = manager::DeviceManager { config };

    let context = Context::new().unwrap();

    match opt {
        Opt { list: true, .. } => {
            list_devices(&context);
        }
        Opt { watch: true, .. } => {
            watch(&context);
        }
        Opt {
            device_name: Some(name),
            ..
        } => {
            describe(&context, &name);
        }
        _ => listen(&manager, &context),
    }
}

/// List input or block devices
fn list_devices(context: &Context) {
    let mut input_enumerator = Enumerator::new(context).unwrap();
    input_enumerator.match_subsystem("input").unwrap();
    for device in input_enumerator.scan_devices().unwrap() {
        println!("Device input\t\t{}", device.sysname().to_string_lossy());
    }

    let mut block_enumerator = Enumerator::new(context).unwrap();
    block_enumerator.match_subsystem("block").unwrap();
    for device in block_enumerator.scan_devices().unwrap() {
        println!("Device block\t\t{}", device.sysname().to_string_lossy());
    }
}

/// Describe device by listing its properties
fn describe(context: &Context, name: &str) {
    let mut enumerator = Enumerator::new(context).unwrap();

    enumerator.match_sysname(name).unwrap();
    for device in enumerator.scan_devices().unwrap() {
        println!("******************************************************************************");
        for property in device.properties() {
            println!(
                "{}: {}",
                property.name().to_string_lossy(),
                property.value().to_string_lossy()
            );
        }
    }
}

/// Listen for changes on plugged devices
fn listen(manager: &manager::DeviceManager, context: &Context) {
    let mut enumerator = Enumerator::new(context).unwrap();

    for device in enumerator.scan_devices().unwrap() {
        manager.handle_device(&device);
    }

    let monitor = Monitor::new(context).unwrap();
    let mut socket = monitor.listen().unwrap();

    loop {
        match socket.receive_event() {
            Some(event) => manager.handle_change(&event),
            None => sleep(Duration::from_millis(SLEEP_DURATION)),
        }
    }
}

/// Watch for changes and display the device properties
fn watch(context: &Context) {
    let monitor = Monitor::new(context).unwrap();
    let mut socket = monitor.listen().unwrap();

    loop {
        match socket.receive_event() {
            Some(event) => describe(context, &event.device().sysname().to_string_lossy()),
            None => sleep(Duration::from_millis(SLEEP_DURATION)),
        }
    }
}

/// Load configuration from file.
/// Create configuration file from defaults if needed.
fn get_config() -> Result<config::Config, std::io::Error> {
    let xdg_dirs = xdg::BaseDirectories::with_prefix(env!("CARGO_PKG_NAME"))?;
    let config_path = xdg_dirs.place_config_file("config.toml")?;

    if !config_path.exists() {
        let toml = toml::to_string(&config::Config::default()).unwrap();
        let mut file = File::create(&config_path)?;
        file.write_all(toml.as_bytes())?;
    }

    config::Config::from_file(config_path)
}


src/manager.rs
use crate::config;
use libudev::{Device, Event, EventType};
use notify_rust::Notification;
use std::process::Command;
use std::collections::HashMap;

pub struct DeviceManager {
    pub config: config::Config,
}

impl DeviceManager {
    pub fn handle_device(&self, device: &Device) {
        for (name, watched) in &self.config.devices {
            if is_matching(&watched.properties, device) {
                notify(&format!("{} plugged", name));
                execute(&watched.on_plugged);
            }
        }
    }

    pub fn handle_change(&self, event: &Event) {
        for (name, watched) in &self.config.devices {
            if is_matching(&watched.properties, event.device()) {
                match event.event_type() {
                    EventType::Add => {
                        notify(&format!("{} plugged", name));
                        execute(&watched.on_plugged);
                    }
                    EventType::Remove => {
                        notify(&format!("{} unplugged", name));
                        execute(&watched.on_unplugged);
                    }
                    _ => break,
                }
            }
        }
    }
}

/// Check if device match with given property set
fn is_matching(properties: &HashMap<String, String>, device: &Device) -> bool {
    let mut matching_properties: usize = 0;
    for property in device.properties() {
        let name = property.name().to_os_string().into_string().expect("Could not convert property name");
        let value = property.value().to_os_string().into_string().expect("Could not convert property value");

        if properties.contains_key(&name) && properties.get(&name).unwrap() == &value {
            matching_properties += 1;
        }
    }

    // Prevent matching on empty property set
    !properties.is_empty() && matching_properties == properties.len()
}

fn execute(command: &str) {
    let cmd: Vec<_> = command.split(' ').collect();
    let run = Command::new(cmd[0])
        .args(&cmd[1..])
        .spawn();
    if run.is_ok() {
        run.unwrap();
    } else {
        eprintln!("Failed to run command: {}", command);
    }
}

fn notify(message: &str) {
    let notification = Notification::new()
        .summary(env!("CARGO_PKG_NAME"))
        .body(message)
        .icon("keyboard")
        .show();
    if notification.is_ok() {
        notification.unwrap();
    } else {
        eprintln!("Unable to notify: {}", message);
    }
}


src/config.rs
use std;
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use serde_derive::{Deserialize, Serialize};
use toml;

#[derive(Deserialize, Serialize)]
pub struct Config {
    pub devices: HashMap<String, WatchedDevice>,
}

#[derive(Deserialize, Serialize)]
pub struct WatchedDevice {
    pub on_plugged: Option<String>,
    pub on_unplugged: Option<String>,
    pub properties: HashMap<String, String>,
}

impl Config {
    pub fn from_file(path: std::path::PathBuf) -> Result<Self, std::io::Error> {
        let mut file = File::open(path)?;
        let mut buffer = String::new();
        file.read_to_string(&mut buffer)?;
        Ok(parse_config(&buffer))
    }
}

fn parse_config(config: &str) -> Config {
    match toml::from_str(config) {
        Ok(config) => config,
        Err(e) => {
            eprintln!("{}", e);
            std::process::exit(1);
        }
    }
}

impl Default for Config {
    fn default() -> Config {
        let mut properties = HashMap::new();
        // Ergodox sample values
        properties.insert("ID_VENDOR_ID".to_owned(), "feed".to_owned());
        properties.insert("ID_MODEL_ID".to_owned(), "1307".to_owned());
        properties.insert("ID_INPUT_KEYBOARD".to_owned(), "1".to_owned());
        properties.insert("KEY".to_owned(), "1000000000007 ff9f207ac14057ff febeffdfffefffff fffffffffffffffe".to_owned());
        let mut devices = HashMap::new();
        devices.insert(
            "ergodox".to_owned(),
            WatchedDevice {
                properties,
                on_plugged: Some("setxkbmap us".to_owned()),
                on_unplugged: Some("setxkbmap fr oss -option ctrl:nocaps".to_owned()),
            },
        );
        Config { devices }
    }
}

#[test]
fn test_optional_cmd() {
    let config = "
    [devices.test]
    on_plugged = \"ok\"
    [devices.test.properties]
    KEY = \"12345\"
    ";
    let parsed = parse_config(config);

    assert!(parsed.devices.contains_key("test"));
    assert!(!parsed.devices["test"].properties.contains_key("on_unplugged"));
}
